DeliveryApi/ApiHandlers/src/com/ilmlf/delivery/api/handlers/PopulateFarmDb.java

```diff
@@ -0,0 +1,148 @@
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.SQLSyntaxErrorException;
+import java.sql.Statement;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.json.JSONObject;
+
+/**
+ * Create required tables and a DB user for Delivery application.
+ *
+ * <p>
+ * CloudFormation Custom Resource will call Lambda function with this code when
+ * the stack is created. This handler will run `dbinit.sql`.
+ * </p>
+ *
+ * <p>
+ * Apart from creating tables, it will create a user in the DB with the credential from
+ * DB_USER_SECRET. API Lambda functions and RDS proxy use this credential to connect to DB.
+ * </p>
+ *
+ * <p>
+ * The Lambda functions will only know the username. They have a Role which allows them
+ * to connect to RDS proxy via this user. (e.g. Action: "rds-db:connect", Resources:
+ * "arn:aws:rds-db:REGION:ACCOUNT_ID:dbuser:*\/DB_USERNAME")
+ * </p>
+ *
+ * <p>
+ * RDS proxy will retrieve the password from DB_USER_SECRET and handle the connection to
+ * the DB instance.
+ * </p>
+ */
+public class PopulateFarmDb implements RequestHandler<Object, String> {
+  /**
+   * Database connection info.
+   */
+  static String DB_ENDPOINT = System.getenv("DB_ENDPOINT");
+  static String DB_REGION = System.getenv("DB_REGION");
+
+  /**
+   * Admin credentials for RDS instance.
+   */
+  static final JSONObject DB_ADMIN_SECRET = SecretsUtil.getSecret(DB_REGION, System.getenv("DB_ADMIN_SECRET"));
+  static final String DB_ADMIN_USER = (String) DB_ADMIN_SECRET.get("username");
+  static final String DB_ADMIN_PWD = (String) DB_ADMIN_SECRET.get("password");
+
+  /**
+   * User credentials for IAM Authorization via RDS Proxy.
+   */
+  static final JSONObject DB_USER_SECRET = SecretsUtil.getSecret(DB_REGION, System.getenv("DB_USER_SECRET"));
+  static final String DB_USERNAME = (String) DB_USER_SECRET.get("username");
+  static final String DB_PASSWORD = (String) DB_USER_SECRET.get("password");
+
+  static Connection con = null;
+  private static final Logger logger = LogManager.getLogger(CreateSlots.class);
+
+  static {
+    try {
+      con = DbUtil.createConnectionViaUserPwd(DB_ADMIN_USER, DB_ADMIN_PWD, DB_ENDPOINT);
+
+    } catch (Exception e) {
+      logger.info("INIT connection FAILED");
+      logger.error(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Entry point for Custom Resource call. This function executes dbinit.sql and
+   * creates a user with the username and password from DB_SECRET.
+   *
+   * @param event Event object from CloudFormation Custom Resource
+   * @param context Context object
+   * @return Result to Custom Resource
+   */
+  @Override
+  public String handleRequest(Object event, Context context) {
+    try {
+      String script = "./com/ilmlf/db/dbinit.sql";
+      FileReader fr = new FileReader(script);
+      BufferedReader br = new BufferedReader(fr);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 103, 108. The resource is closed at line 121. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).

Reply:
switched to try-with-resources for FileReader, BufferedReader, and Statement closing